File: WEEK4/bj5_1516.py
import sys
from collections import deque

input = sys.stdin.readline
n = int(input())

adjList = [[] for _ in range(n + 1)]
inDegree = [0] * (n + 1)
buildTime = [0] * (n + 1)
resultTime = [0] * (n + 1)

for i in range(1, n + 1):
    line = list(map(int, input().split()))
    buildTime[i] = line[0]
    resultTime[i] = line[0]
    for j in range(1, len(line) - 1):
        pre = line[j]
        adjList[pre].append(i)
        inDegree[i] += 1

queue = deque()
for i in range(1, n + 1):
    if inDegree[i] == 0:
        queue.append(i)

while queue:
    current = queue.popleft()
    for next_node in adjList[current]:
        resultTime[next_node] = max(
            resultTime[next_node], resultTime[current] + buildTime[next_node]
        )
        inDegree[next_node] -= 1
        if inDegree[next_node] == 0:
            queue.append(next_node)

for i in range(1, n + 1):
    print(resultTime[i])

File: WEEK4/bj5_1516.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const adjList = Array.from({ length: n + 1 }, () => []);
const inDegree = new Array(n + 1).fill(0);
const buildTime = new Array(n + 1).fill(0);
const resultTime = new Array(n + 1).fill(0);

for (let i = 1; i <= n; i++) {
    const line = input[i].split(' ').map(Number);
    buildTime[i] = line[0];
    resultTime[i] = line[0];

    for (let j = 1; j < line.length - 1; j++) {
        const pre = line[j];
        adjList[pre].push(i);
        inDegree[i]++;
    }
}

const queue = [];
let head = 0;
for (let i = 1; i <= n; i++) {
    if (inDegree[i] === 0) {
        queue.push(i);
    }
}

while (head < queue.length) {
    const current = queue[head++];

    for (const next of adjList[current]) {
        resultTime[next] = Math.max(resultTime[next], resultTime[current] + buildTime[next]);
        inDegree[next]--;

        if (inDegree[next] === 0) {
            queue.push(next);
        }
    }
}

let result = '';
for (let i = 1; i <= n; i++) {
    result += resultTime[i] + '\n';
}
console.log(result.trim());
